src/Shared/FileMatcher.cs

```diff
@@ -164,17 +169,39 @@ private static string[] GetAccessibleFilesAndDirectories(string path, string pat
                     // do nothing
                 }
             }
+            return Enumerable.Empty<string>();
+        }
 
-            if (entries == null)
+        /// <summary>
+        /// Determine if the given search pattern will match loosely on Windows
+        /// </summary>
+        /// <param name="searchPattern">The search pattern to check</param>
+        /// <returns></returns>
+        private static bool ShouldEnforceMatching(string searchPattern)
+        {
+            if (!NativeMethodsShared.IsWindows)
             {
-                entries = Array.Empty<string>();
+                return false;
             }
-
-            return entries;
+            // NOTE: Windows matches loosely in three cases (in the absence of the * wildcard in the extension):
+            // 1) if the extension ends with the ? wildcard, it matches files with shorter extensions also e.g. "file.tx?" would
+            //    match both "file.txt" and "file.tx"
+            // 2) if the extension is three characters, and the filename contains the * wildcard, it matches files with longer
+            //    extensions that start with the same three characters e.g. "*.htm" would match both "file.htm" and "file.html"
+            // 3) if the ? wildcard is to the left of a period, it matches files with shorter name e.g. ???.txt would match
+            //    foo.txt, fo.txt and also f.txt
+            string extensionPart = Path.GetExtension(searchPattern);
```

Comment:
Please don't allocate here unless needed, make sure this occurs only when if the search for "?." succeeds.

Reply:
[Fixed](https://github.com/maca88/msbuild/blob/b5574289426c99e9567a43fe17d4731e1f4f5380/src/Shared/FileMatcher.cs#L259)